usr/src/uts/common/fs/zfs/range_tree.c

```diff
@@ -33,6 +33,50 @@
 #include <sys/zio.h>
 #include <sys/range_tree.h>
 
+/*
+ * Range trees are tree-based data structures that can be used to
+ * track free space or generally any space allocation information.
+ * A range tree keeps track of individual segments and automatically
+ * provides facilities such as adjacent extent merging and extent
+ * splitting in response to range add/remove requests.
+ *
+ * A range tree starts out completely empty, with no segments in it.
+ * Adding an allocation via range_tree_add to the range tree can either:
+ * 1) create a new extent
+ * 2) extend an adjacent extent
+ * 3) merge two adjacent extents
+ * Conversely, removing an allocation via range_tree_remove can:
+ * 1) completely remove an extent
+ * 2) shorten an extent (if the allocation was near one of its ends)
+ * 3) split an extent into two extents, in effect punching a hole
+ *
+ * A range tree is also capable of 'bridging' gaps when adding
+ * allocations. This is useful for cases when close proximity of
+ * allocations is an important detail that needs to be represented
+ * in the range tree. See range_tree_set_gap(). The default behavior
+ * is not to bridge gaps (i.e. the maximum allowed gap size is 0).
+ *
+ * In order to traverse a range tree, use either the range_tree_walk
+ * or range_tree_vacate functions.
+ *
+ * To obtain more accurate information on individual segment
+ * operations that the range tree performs "under the hood", you can
+ * specify a set of callbacks by passing a range_tree_ops_t structure
+ * to the range_tree_create and range_tree_create_custom functions.
+ * Any callbacks that are non-NULL are then called at the appropriate
+ * times.
+ *
+ * It is possible to store additional custom information with each
+ * segment. This is typically useful when using one or more of the
+ * operation callbacks mentioned above. To do so, use the
+ * range_tree_create_custom function to create your range tree and
+ * pass a custom kmem cache allocator. This allocator must allocate
+ * objects of at least sizeof (range_seg_t). The range tree will use
+ * the start of that object as a range_seg_t to keep its internal
+ * data structures and you can use the remainder of the object to
+ * store arbitrary additional fields as necessary.
```

Comment:
This will be removed with the integration of ss_fill tracking into range_tree itself.